### PR TITLE
[RFR]  Regex for allowing slashes in `PUT file/{uuid}`.  Partially addresses #1885 .

### PR DESCRIPTION
**Pull request by DailyDreaming**
Partially addresses #1885 .  Formal ticket is #1900.

Changes the current `PUT file/{uuid}` name field to allow slashes, but not allow `\.`, `.\`, exactly `.`, exactly `..`, or for the path to lead with `/` (disallowing absolute paths).  See the tests for good and bad paths (which I'd like feedback on!).

Backslashes (Windows file paths) will be disallowed for upload.  Support is intended to occur during download via the CLI, so uploading from windows will need to convert `\` to `/`.

### CODE REVIEW

File: dss-api.yml [hunk 1]
```diff
@@ -1385,7 +1385,7 @@ paths:
                     name:
                       description: Name of the file.
                       type: string
-                      pattern: "^[^/\\\\]+$"
+                      pattern: "(?=^((?!\\.\\/).)*$)(?=^((?!\\/\\.).)*$)(?=^(?!\\/).*$)(?=^(?!~\\/).*$)(?=^(?!\\.\\.$).*)(?=^(?!\\.$).*)"
```
Comment (Bento007) [inline dss-api.yml]: I suggest this simplified regex
```suggestion
                      pattern: "^(\\.?[\\w\\d][\\w\\d\\s.]*)+((\\.?[\\w\\d][\\w\\d\\s.]*)|\\/)*$"
```

File: tests/test_api.py [hunk 2]
```diff
@@ -69,6 +69,60 @@ def test_get_version(self):
         res = self.assertGetResponse("/version", requests.codes.ok, headers=get_auth_header())
         self.assertEquals(res.json['version_info']['version'], os.environ['DSS_VERSION'])
 
+    @testmode.standalone
+    def test_put_file_pattern(self):
```
Comment (Bento007) [inline tests/test_api.py]: These tests belong in `test_file.py`

File: tests/test_api.py [hunk 3]
```diff
@@ -69,6 +69,60 @@ def test_get_version(self):
...
+        """
+        Tests the regex pattern filtering on the PUT file/{uuid} endpoint.
+
+        Ensures that paths with leading slashes (absolute paths) and
+        relative path shortcuts (".", "~/", and "..") are disallowed.
+        """
+
+        examples_of_bad_paths = ['.', '..',
```
Comment (Bento007) [inline tests/test_api.py]: are `\\` or `//` allowed, and should `good..path` be allowed?

File: tests/test_api.py [hunk 4]
```diff
@@ -69,6 +69,60 @@ def test_get_version(self):
...
+                                 '~/path',
+                                 './', '../', './path', './../path', '../path', '../../path',
+                                 'path/../path', 'path/..',
+                                 '/path', '/path.json', '/bad..path']
+
+        examples_of_good_paths = ['path', 'path.json', 'good..path', 'path.json/', 'good..path/',
+                                  'pa/th.json', 'go/od..path', 'a/b/c/d', 'a/b.c/d', '.bashrc']
+
+        for path, replica in [("files", "aws"), ("files", "gcp")]:
```
Comment (Bento007) [inline tests/test_api.py]: Only test one replica. The replica does not affect the outcome.

File: tests/test_api.py [hunk 5]
```diff
@@ -69,6 +69,60 @@ def test_get_version(self):
...
+            es_query={},
+            callback_url="https://take.me.to.funkytown",
+            source_url="s3://urlurlurlbaby",
+            description="supercalifragilisticexpialidocious",
+            details={},
+            reason="none",
+            contents=[],
+            name="frank",
+        )
+        put_url = str(UrlBuilder().set(path=f"/v1/{path}/{uuid}")
```
Comment (Bento007) [inline tests/test_api.py]: should be
```suggestion
        put_url = str(UrlBuilder().set(path=f"/v1/files/{uuid}")
```

File: tests/test_api.py [hunk 6]
```diff
@@ -69,6 +69,60 @@ def test_get_version(self):
...
+                                 '~/path',
+                                 './', '../', './path', './../path', '../path', '../../path',
+                                 'path/../path', 'path/..',
+                                 '/path', '/path.json', '/bad..path']
+
+        examples_of_good_paths = ['path', 'path.json', 'good..path', 'path.json/', 'good..path/',
+                                  'pa/th.json', 'go/od..path', 'a/b/c/d', 'a/b.c/d', '.bashrc']
+
+        for path, replica in [("files", "aws"), ("files", "gcp")]:
+            with self.subTest(path=path, replica=replica):
```
Comment (Bento007) [inline tests/test_api.py]: The subtest should be something like this
``` python
        for bad_path in examples_of_bad_paths:
            with self.subTest(path=bad_path, replica=replica):
                    self.put_file_reponse(bad_path, replica, expected_code=requests.codes.bad_request)
```

Do the samething for good_path

File: tests/test_api.py [hunk 7]
```diff
@@ -69,6 +69,76 @@ def test_get_version(self):
...
+
+        for path, replica in [("bundles", "aws"), ("bundles", "gcp")]:
+            with self.subTest(path=path, replica=replica):
+                for bad_path in examples_of_bad_paths:
+                    self.put_bundles_reponse(bad_path, replica, expected_code=[requests.codes.bad_request])
+                for good_path in examples_of_good_paths:
+                    self.put_bundles_reponse(good_path, replica, expected_code=[requests.codes.ok, requests.codes.created])
+
+    def get_test_fixture_bucket(self, replica: Replica) -> str:
+        if replica == 'aws':
```
Comment (Bento007) [inline tests/test_api.py]: use `Replica.aws.name`

File: dss-api.yml (hunk 1 above)
Comment (DailyDreaming) [reply dss-api.yml]: Sure, sounds good.

File: tests/test_api.py (hunk 2 above)
Comment (DailyDreaming) [reply tests/test_api.py]: I was asked to put them here, though I'm not sure if it was a strong suggestion.  I'll bring it up over slack.

File: tests/test_api.py (hunk 3 above)
Comment (DailyDreaming) [reply tests/test_api.py]: `//` is allowed on a normal Linux system.  `good..path` is weird but valid on most sytems I believe.  We should all talk about what and should not be allowed on Monday.

File: tests/test_api.py (hunk 4 above)
Comment (DailyDreaming) [reply tests/test_api.py]: Alright.

File: tests/test_api.py (hunk 6 above)
Comment (DailyDreaming) [reply tests/test_api.py]: Why?

File: tests/test_api.py (hunk 7 above)
Comment (DailyDreaming) [reply tests/test_api.py]: Done.

File: tests/test_api.py (hunk 6 above)
Comment (DailyDreaming) [reply tests/test_api.py]: Thanks for the catch.  Looking back, I don't know why I did it that way.

File: tests/test_api.py (hunk 2 above)
Comment (DailyDreaming) [reply tests/test_api.py]: `test_file.py` works!  Migrated the tests over.

File: dss-api.yml [hunk 8]
```diff
@@ -1385,7 +1385,7 @@ paths:
                     name:
                       description: Name of the file.
```
Comment (xbrianh) [inline dss-api.yml]: Consider briefly summarizing usage of slashes here.

Comment (DailyDreaming) [reply dss-api.yml]: Good call.  Adding it in.

**Comment by DailyDreaming** [comment]
Flaky test:
```
======================================================================
FAIL: test_bundle_delete (__main__.TestBundleApi) [gcp, True]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_bundle.py", line 576, in test_bundle_delete
    self._test_bundle_delete(*test)
  File "tests/test_bundle.py", line 621, in _test_bundle_delete
    self.delete_bundle(replica, bundle_uuid, bundle_version, authorized=authorized)
  File "tests/test_bundle.py", line 778, in delete_bundle
    headers=get_auth_header(authorized=authorized),
  File "/home/travis/build/HumanCellAtlas/data-store/tests/infra/assert_mixin.py", line 99, in assertResponse
    self.assertEqual(response.status_code, expected_code)
AssertionError: 500 != 200
```

**Comment by DailyDreaming** [comment]
@Bento007 Have I addressed all issues?

File: dss-api.yml [hunk 9]
```diff
@@ -1383,9 +1383,16 @@ paths:
...
                     name:
-                      description: Name of the file.
+                      description: >
+                        Name of the file.
+
+                        - Absolute paths (beginning with a slash) are not allowed.
+                        - Relative paths must not have unix shortcuts (i.e. ".", "..", "~/")
+                        - No "\" (back)slashes allowed.  e.g. "C:\\path" or "path\file.json"
+                        - Only relative paths with "/" (forward)slashes allowed.  e.g. "path/file.json" or "file.json"
+                        - Please upload using "/" (forward)slashes, even if on Windows.  DL will convert back.
```
Comment (Bento007) [inline dss-api.yml]: expand DL to downloads or remove last sentence.

Comment (Bento007) [reply dss-api.yml]: The API will not modify the slashes in get file or bundle. This would be done in the CLI.

Comment (DailyDreaming) [reply dss-api.yml]: I'll remove the last sentence.  I realize it was discussed offline and had planned to make the corresponding changes in `dcp-cli`, but that that should be a separate PR.  I'll create an issue for it.

**Pull request merged**